development/extensions/tutorial_testing.rst

```diff
@@ -0,0 +1,913 @@
+config influences the ``\phpbb\user::lang()`` call:
+
+.. code-block:: php
+
+    $l_message = empty($this->config['acme_demo_goodbye']) ? 'DEMO_HELLO' : 'DEMO_GOODBYE';
+    $this->template->assign_var('DEMO_MESSAGE', $this->user->lang($l_message, $name));
+
+This is what we check with the third argument ``$expected_language`` of our test
+method:
+
+.. code-block:: php
+
+            $this->user->expects($this->once())
+                ->method('lang')
+                ->with($expected_language, $name)
+                ->willReturn($language_return);
+
+For more information about Mocks and phpunit, please have a look at the
+`phpunit Documentation <https://phpunit.de/manual/current/en/test-doubles.html>`_.
+
+phpunit configuration file
+--------------------------
+
+Before we can run the tests we need to create the configuration file for
+phpunit. Don't be scared if you do not understand it, you should not need to
+edit anything in that file.
+The file should be stored as ``ext/acme/demo/phpunit.xml.dist``:
+
+.. code-block:: xml
+
+    <?xml version="1.0" encoding="UTF-8"?>
+    <phpunit backupGlobals="true"
+             backupStaticAttributes="true"
+             colors="true"
+             convertErrorsToExceptions="true"
+             convertNoticesToExceptions="true"
+             convertWarningsToExceptions="true"
+             processIsolation="false"
+             stopOnFailure="false"
+             syntaxCheck="false"
+             verbose="true"
+             bootstrap="../../../../tests/bootstrap.php"
+    >
+        <testsuites>
+            <testsuite name="Extension Test Suite">
+                <directory suffix="_test.php">./tests</directory>
+                <exclude>./tests/functional</exclude>
+            </testsuite>
+            <testsuite name="Extension Functional Tests">
+                <directory suffix="_test.php" phpVersion="5.3.19" phpVersionOperator=">=">./tests/functional/</directory>
+            </testsuite>
+        </testsuites>
+
+        <filter>
+            <blacklist>
+                <directory>./tests/</directory>
+            </blacklist>
+            <whitelist processUncoveredFilesFromWhitelist="true">
+                <directory suffix=".php">./</directory>
+                <exclude>
+                    <directory suffix=".php">./language/</directory>
+                    <directory suffix=".php">./migrations/</directory>
+                    <directory suffix=".php">./tests/</directory>
+                </exclude>
+            </whitelist>
+        </filter>
+    </phpunit>
+
+Now we can finally run the tests by executing the following command::
+
+    $ ./phpBB/vendor/bin/phpunit -c phpBB/ext/acme/demo/phpunit.xml.dist
+    PHPUnit 4.1.0 by Sebastian Bergmann.
+
+    Configuration read from /home/nickv/phpBB/Ascraeus/phpBB/ext/acme/demo/phpunit.xml.dist
+
+    ...
+
+    Time: 101 ms, Memory: 9.00Mb
+
+    OK (3 tests, 12 assertions)
+
```

Comment:
Add a line that individual tests can be run by doing:
`$ ./phpBB/vendor/bin/phpunit phpBB/ext/acme/demo/tests/controller/main_test.php`